File: backend/youthconnect_id/src/main/java/com/youthconnect/youthconnect_id/services/ConcernService.java
package com.youthconnect.youthconnect_id.services;

import java.util.List;

import com.youthconnect.youthconnect_id.dto.AdminConcernUpdateRequest;
import com.youthconnect.youthconnect_id.dto.ConcernRequest;
import com.youthconnect.youthconnect_id.dto.ConcernResponse;
import com.youthconnect.youthconnect_id.dto.ConcernUpdateRequest;
import com.youthconnect.youthconnect_id.enums.ConcernStatus;

public interface ConcernService {
    // Youth
    ConcernResponse submitConcern(ConcernRequest request);
    List<ConcernResponse> getOwnConcerns(int youthId);
    ConcernResponse getConcernById(int concernId);
    ConcernResponse editConcern(int concernId, ConcernUpdateRequest request);
    void cancelConcern(int concernId);
    void deleteConcern(int concernId);

    // Admin
    List<ConcernResponse> getAllConcerns();
    ConcernResponse updateConcernStatus(int concernId, ConcernStatus status);
    void addConcernUpdate(int concernId, AdminConcernUpdateRequest request);
}

File: frontend/youthconnect-frontend/src/app/services/concern.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface ConcernRequest {
    youthId: number;
    typeOfConcern: 'PROJECT_CONCERN' | 'COMMUNITY_CONCERN' | 'SYSTEM_CONCERN';
    title: string;
    description: string;
}

export interface ConcernUpdateRequest {
    typeOfConcern: 'PROJECT_CONCERN' | 'COMMUNITY_CONCERN' | 'SYSTEM_CONCERN';
    title: string;
    description: string;
}

export interface ConcernResponse {
    concernId: number;
    youthId: number;
    typeOfConcern: 'PROJECT_CONCERN' | 'COMMUNITY_CONCERN' | 'SYSTEM_CONCERN';
    title: string;
    description: string;
    status: 'OPEN' | 'IN_PROGRESS' | 'RESOLVED' | 'CLOSED';
    createdAt: string;
    updatedAt?: string;
}

@Injectable({
    providedIn: 'root'
})
export class ConcernService {
    private http = inject(HttpClient);
    private apiUrl = 'http://localhost:8080/api/concerns';

    submitConcern(request: ConcernRequest): Observable<ConcernResponse> {
        return this.http.post<ConcernResponse>(this.apiUrl, request);
    }

    getOwnConcerns(youthId: number): Observable<ConcernResponse[]> {
        return this.http.get<ConcernResponse[]>(`${this.apiUrl}/youth/${youthId}`);
    }

    getConcernById(concernId: number): Observable<ConcernResponse> {
        return this.http.get<ConcernResponse>(`${this.apiUrl}/${concernId}`);
    }

    editConcern(concernId: number, request: ConcernUpdateRequest): Observable<ConcernResponse> {
        return this.http.put<ConcernResponse>(`${this.apiUrl}/${concernId}`, request);
    }

    deleteConcern(concernId: number): Observable<any> {
        return this.http.delete(`${this.apiUrl}/${concernId}`, { responseType: 'text' });
    }

    cancelConcern(concernId: number): Observable<any> {
        return this.http.patch(`${this.apiUrl}/${concernId}/cancel`, {}, { responseType: 'text' });
    }

    getAllConcernsForSkOfficial(): Observable<ConcernResponse[]> {
        return this.http.get<ConcernResponse[]>(`${this.apiUrl}`);
    }
}
